### ISSUE TITLE
search messages in room

### ISSUE THREAD

**Issue by KiaraGrouwstra** [element-hq/element-x-android#3709]
#### What would you like to do?

search the messages in a room

#### Why would you like to do it?

faster than reading all messages

#### How would you like to achieve it?

have a search bar in the room menu

### Have you considered any alternatives?

use a different client

### Are you willing to provide a PR?

No

**Comment by jdauphant** [comment]
If search on encrypted all messages are technically complicated : in that case, search index could be limited to X last days (30 days for example and this setting could be change in settings).

**Comment by frandavid100** [comment]
> If search on encrypted all messages are technically complicated : in that case, search index could be limited to X last days (30 days for example and this setting could be change in settings).

Maybe search whatever messages are stored locally, then have a "load previous results" that stores more messages and searches those?

**Comment by jdauphant** [comment]
> > If search on encrypted all messages are technically complicated : in that case, search index could be limited to X last days (30 days for example and this setting could be change in settings).
> 
> Maybe search whatever messages are stored locally, then have a "load previous results" that stores more messages and searches those?

"Load previous results" could work if there is no performance issues.
Also, to be sure the user get how the search work and appreciate it, I think we should search in all messages for a period (with no gap) and we should tell him this period ("search on the last 15 days").

**Comment by frandavid100** [comment]
> "Load previous results" could work if there is no performance issues.

Is there any reason to think there would?

> Also, to be sure the user get how the search work and appreciate it, I think we should search in all messages for a period (with no gap) and we should tell him this period ("search on the last 15 days").

Not sure about that, since it might complicate the UI and the devs seem to be going in the opposite direction; but I'm not against it either.

**Comment by lionking102** [comment]
I also encountered this issue. A search function would be amazing and, I believe, quite important for many users.

How do other major messaging apps handle searching through the entire chat history?

**Comment by maurerle** [comment]
This issue is very important for me, as I would like to search the chat history from time to time.
Having this feature on the desktop element client is a good alternative, but somehow not really sufficient :/

**Comment by saihaze** [comment]
When implementing this, CJK should be considered. Tokenising these languages is quite different because they are not space-seperated.

**Comment by nathanael-h** [comment]
Related issue for iOS: https://github.com/element-hq/element-x-ios/issues/3766

**Comment by vanyasem** [comment]
Meta Issue: https://github.com/element-hq/element-meta/issues/1044
Related Meta Issue: https://github.com/element-hq/element-meta/issues/2004

Related Element Web issue: https://github.com/element-hq/element-web/issues/16483

Related issue for old Android Element: https://github.com/element-hq/element-android/issues/5268
Related issue for old iOS Element: https://github.com/element-hq/element-ios/issues/7377

Somewhat related issue for Desktop Element concerning `seshat` imperfections: https://github.com/element-hq/element-web/issues/32356

Discussion about server-side search for encrypted rooms: https://github.com/element-hq/element-web/issues/4982#issuecomment-602249028

**Comment by oldu9412** [comment]
Yes, it's very important, it's a blocking issue for moving to Element X. Many people won't want to come to Matrix/Element and will prefer to go to Discord, because the search function is very important to them
